imxrt1060-hal/src/trng.rs

```diff
@@ -0,0 +1,402 @@
+        self.0.read(dest).map_err(|e| {
+            let code = if e.0.bits == 0 { 1 << 31 } else { e.0.bits };
+            // Safety: Highest bit is set above if no flags were set.
+            unsafe { core::num::NonZeroU32::new_unchecked(code).into() }
+        })
+    }
+}
+
+impl Read for TRNG {
+    type Error = Error;
+    // e-h RNG Read is a *blocking* trait, so no WouldBlock here
+    // Read is part of the unproven API and will likely be changed in the future
+
+    fn read(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        let mut data = [0; 4];
+        let mut index = 4;
+        for b in buffer.iter_mut() {
+            if index == 4 {
+                data = nb::block!(self.next_u32())?.to_be_bytes();
+                index = 0;
+            }
+            *b = data[index];
+            index += 1;
+        }
+        Ok(())
+    }
+}
+
+/// A TRNG error occurred, such as a statistical test failing.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct Error(pub ErrorFlags);
+
+bitflags::bitflags! {
+    /// Specific errors that may occur during entropy generation
+    pub struct ErrorFlags : u32 {
+        // STATUS register starts here (automatically set from bits)
+        /// 1-bit run sampling 0s test failed
+        const TF1BR0 = 1 << 0;
+        /// 1-bit run sampling 1s test failed
+        const TF1BR1 = 1 << 1;
+        /// 2-bit run sampling 0s test failed
+        const TF2BR0 = 1 << 2;
+        /// 2-bit run sampling 1s test failed
+        const TF2BR1 = 1 << 3;
+        /// 3-bit run sampling 0s test failed
+        const TF3BR0 = 1 << 4;
+        /// 3-bit run sampling 1s test failed
+        const TF3BR1 = 1 << 5;
+        /// 4-bit run sampling 0s test failed
+        const TF4BR0 = 1 << 6;
+        /// 4-bit run sampling 1s test failed
+        const TF4BR1 = 1 << 7;
+        /// 5-bit run sampling 0s test failed
+        const TF5BR0 = 1 << 8;
+        /// 5-bit run sampling 1s test failed
+        const TF5BR1 = 1 << 9;
+        /// 6-plus-bit run sampling 0s test failed
+        const TF6PBR0 = 1 << 10;
+        /// 6-plus-bit run sampling 1s test failed
+        const TF6PBR1 = 1 << 11;
+        /// Sparse bit test failed
+        const TFSB = 1 << 12;
+        /// Long run test failed
+        const TFLR = 1 << 13;
+        /// Poker test failed
+        const TFP = 1 << 14;
+        /// Mono bit test failed
+        const TFMB = 1 << 15;
+        // MCTL register starts here (set manually)
+        /// Count taken during entropy generation was outside the defined range of FRQ_MIN to FRQ_MAX
+        const FCT_FAIL = 1 << 16;
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "An error occurred in the TRNG module")
+    }
+}
+
+/// The specified retry count was outside of the valid range of `1..=15`.
+#[derive(Copy, Clone, Debug)]
+pub struct InvalidRetryCountError;
```

Comment:
This error struct could be constructed by anyone outside of the HAL crate. That's not a big deal, but we might want to give others the guarantee that we're the only people that can create this.

```suggestion
pub struct InvalidRetryCountError(());
```

Could demonstrate this with a failing doctest:

```rust
/// ```compile_fail
/// use imxrt1060_hal as hal;
/// use hal::trng::InvalidRetryCountError;
/// let err = InvalidRetryCountError(());
/// ```
#[cfg(doctest)]
struct UnconstructibleError;
```

Reply:
Thanks for catching that! Wasn't paying enough attention.

The config errors of I2C and SPI (`ClockSpeedError`, `PinLowTimeoutError`, `BusIdleTimeoutError`, `ModeError`) are also missing the private empty tuple. (I think I just mimicked them without using my brain.) Will make a PR to also make those unconstructible.

Added a slightly modified version of that test. With the `(())` suffix it passed on both the wrongly constructible error and the correct private error.